### ISSUE TITLE
Generated SQL incorrect when grouping by expression and ordering by same expression 2.2.6

### ISSUE THREAD

**Issue by lmkz** [dotnet/efcore#21442]
Bug in EFCore 2.2.6; generated SQL not correct. When query contains group by based on an expression, and that expression is used to order the results, SQL fails. See examples below.

### Steps to reproduce

Create a table in either SQL Server or Postgres:

``` SQL
CREATE TABLE [dbo].[CourseSerial](
	[SerialID] [uniqueidentifier] NOT NULL,
	[CourseID] [uniqueidentifier] NULL,
	[CourseCode] [varchar](15) NOT NULL,
	[CourseName] [varchar](40) NOT NULL,
	[SerialCode] [varchar](50) NULL,
	[SerialName] [varchar](max) NULL)
```

``` C#
 public string DoTestGroup(bool groupByBatch = false) {
            var now = DateTime.Now.ToString();

            var query = from s in cx.CourseSerial
                        group 
                            s.CourseName                        
                         by new {
                             s.CourseName,
                             BatchDate = groupByBatch ? s.SerialCode : now
                         } into eventGroup
                        orderby eventGroup.Key.BatchDate
                        select new {
                            EventType = eventGroup.Key.CourseName
                        };

            var list = query.ToList();

            return "OK";
        }
```

SQL executed:
Microsoft.EntityFrameworkCore.Database.Command:Error: Failed executing DbCommand (238ms) [Parameters=[@__groupByBatch_0='False', @__now_1='29/06/2020 8:59:17 AM' (Size = 4000)], CommandType='Text', CommandTimeout='30']
SELECT [s].[CourseName] AS [EventType]
FROM [CourseSerial] AS [s]
GROUP BY [s].[CourseName], CASE
    WHEN @__groupByBatch_0 = 1
    THEN [s].[SerialCode] ELSE @__now_1
END
ORDER BY [BatchDate]

System.Data.SqlClient.SqlException (0x80131904): Invalid column name 'BatchDate'.

Workaround: Add expression field to select list.

### Further technical details

EF Core version: 2.2.6
Database provider: Microsoft.EntityFrameworkCore.SqlServer or postgres
Target framework: .NET Core 2.1
Operating system: Windows 10
IDE: Visual Studio 2019 16.6.2

**Comment by lmkz** [comment]
Note works in EF Core 3+

**Comment by ajcvickers** [comment]
@lmkz EF Core 2.2 is [out-of-support](https://docs.microsoft.com/ef/core/what-is-new/). Please update to 3.1.

**Comment by lmkz** [comment]
> @lmkz EF Core 2.2 is [out-of-support](https://docs.microsoft.com/ef/core/what-is-new/). Please update to 3.1.

OK what about 2.1? Presumably it has the same bug??

**Comment by ajcvickers** [comment]
@lmkz Presumably, but the bar for fixing issues in 2.1 is extremely high. It is very unlikely that this will be fixed in 2.1.

**Comment by lmkz** [comment]
> @lmkz Presumably, but the bar for fixing issues in 2.1 is extremely high. It is very unlikely that this will be fixed in 2.1.

Roger that, thanks for prompt reply.
